### PR TITLE
Dev/kind loop

### PR DESCRIPTION
**Pull request by droot**
Adds a kind-based environment for exercising the DRA workload scheduler against
real GPUs on a single Linux GPU VM, so claim allocation and worker scheduling
can be tested without a GKE cluster. Cluster teardown and rebuild takes about a
minute, which makes claim behaviour cheap to test from a clean slate.

`dev/kind/` holds the scripts and `k8s/deploy/kind-dra/` overlays the GKE
manifests with local image names, an RWO local-path PVC, and the GKE-only
pieces removed.

## Using it

Everything talks to the local docker daemon and kube context, so it runs on the
GPU VM. Get your tree there the way the rest of the repo does:

```bash
make push-vm REMOTE_HOST=<host>                  # from the workstation

Then, on the VM (or ssh <host> 'cd ~/open-rl && make ...' without logging in):

┌───────────────────────────────┬─────────────────────────────────────────────────────────────────┐
│                               │                                                                 │
├───────────────────────────────┼─────────────────────────────────────────────────────────────────┤
│ make kind-host-setup          │ one-time: docker, nvidia-container-toolkit, kind, kubectl, helm │
├───────────────────────────────┼─────────────────────────────────────────────────────────────────┤
│ make kind-create              │ cluster, DRA driver, registry                                   │
├───────────────────────────────┼─────────────────────────────────────────────────────────────────┤
│ make kind-deploy              │ build, publish, apply                                           │
├───────────────────────────────┼─────────────────────────────────────────────────────────────────┤
│ make kind-gateway             │ rebuild and roll just the slim gateway — the fast inner loop    │
├───────────────────────────────┼─────────────────────────────────────────────────────────────────┤
│ make kind-client              │ rebuild the e2e client image                                    │
├───────────────────────────────┼─────────────────────────────────────────────────────────────────┤
│ make kind-e2e                 │ the cluster-e2e harness against the kind context                │
├───────────────────────────────┼─────────────────────────────────────────────────────────────────┤
│ make kind-status / kind-logs  │ inspect                                                         │
├───────────────────────────────┼─────────────────────────────────────────────────────────────────┤
│ make kind-prune / kind-delete │ reclaim disk / tear down                                        │
└───────────────────────────────┴─────────────────────────────────────────────────────────────────┘

```

## Notes for reviewers

Three parts of the setup are non-obvious and are documented in
dev/kind/README.md:

- A GPU is injected twice — host into the kind node container, then node into
  the pod. host-setup.sh makes nvidia the default docker runtime because
  kind offers no way to pass --runtime.
- The node is labelled nvidia.com/gpu.present=true by hand. The DRA
  kubelet-plugin DaemonSet has a required nodeAffinity on a Node Feature
  Discovery label, and NFD is absent here. Without the label the DaemonSet sits
  at desired: 0 while DeviceClasses still register, so the cluster looks
  healthy and every claim silently stays Pending.
- The node image must be 1.34+, where resource.k8s.io/v1 reached GA. On 1.33
  the API server serves v1beta1 and every claim call 404s.

Images go to a cluster-local registry:2 on the kind docker network rather than
kind load docker-image, which has no layer-level dedup and re-ships the whole
multi-GB server image for a one-line change. The node reaches it through a
containerd mirror, so a single reference resolves from both host and node. That
mirror is applied at cluster creation — an existing cluster must be recreated to
pick it up. :kind-dev is mutable and republished every iteration, hence
imagePullPolicy: Always throughout the overlay.

kind-prune rebuilds the registry and re-pushes rather than running registry garbage-collect: distribution 2.8 does not traverse OCI image indexes, and
collecting deletes the child manifests under a tagged index, leaving a registry
that answers 200 on the tag and 404 on every layer beneath it. Measured on one
box, pruning reclaimed 12.4 GB (24.3 GB → 11.9 GB) in under a minute. The
BuildKit cache is deliberately left alone.

Two changes fall outside dev/kind/ and k8s/deploy/kind-dra/:

- src/server/k8s_worker_manager.py — the gateway now stamps worker
  imagePullPolicy alongside the image it already overrode at render time,
  because worker pod templates live inside ConfigMaps as YAML strings that no
  kustomize patch can reach.
- scripts/run_cluster_e2e.py and k8s/eval/e2e-client-job.yaml — the
  hardcoded imagePullPolicy becomes a placeholder driven by
  --image-pull-policy, defaulting to Always so GKE behaviour is unchanged.

## Testing

Unit tests and lint pass. Exercised on a two-L4 VM: make kind-deploy from a
clean rebuild, then make kind-e2e E2E_SCENARIO=tiny-rl — trainer and sampler
launched against real DRA claims, two steps trained with loss dropping and
mean_reward 0.62 → 0.75, checkpoint written to the shared PVC, workers
cleaned up. kind-status and kind-prune exercised the same way.

LoRA — tiny-rl, clean pass on the branch as-is
[tiny-rl] step=01/2 loss=0.026074 mean_reward=0.62 datums=8
[tiny-rl] step=02/2 loss=-0.026456 mean_reward=0.75 datums=8
final_state_path=/mnt/shared/open-rl/checkpoints/d3038c1b-.../weights/tiny-rl-final
Exit 0, claims workload-type: lora / 24gb, workers cleaned up.

FFT — fft-gsm8k-rl, pass
env/all/reward/total   0.125
Saved checkpoints: {'state_path': '/mnt/shared/open-rl/checkpoints/06f113e8-.../weights/final'}
Training completed successfully          EXIT=0
Claims came out workload-type: full / 24gb — genuinely the full fine-tuning path, trainer and sampler both scheduled against real DRA claims.

Everything ran the new way: make push-vm REMOTE_HOST=l4dev, then ssh l4dev 'cd ~/open-rl && make kind-deploy | kind-gateway | kind-e2e | kind-status | kind-prune'. kind-deploy did a full server-image rebuild from this tree first, so the images under test came from the branch.

Two caveats you need for the review, both of which cost me a false start:

1. tiny-fft-rl does not test FFT. I ran it first and it passed — but the claims came back workload-type: lora. Both tiny examples call create_lora_training_client (examples/tiny/tiny_rl.py:88), and run_tiny only changes the learning rate when the scenario name contains "fft". So the whole tiny-fft* family is LoRA underneath. That's pre-existing and unrelated to this PR, but the names are actively misleading. That's why the real FFT coverage above uses fft-gsm8k-rl.
2. The FFT run required a patch that isn't in this PR. On this branch alone, any full fine-tune stops at No cluster ResourceSlice found matching VRAM memory tier '80gb', because the size-aware tiering fix lives on the merge-prep branch. I applied it to the box's working copy, rebuilt only the gateway, ran the test, then reverted. So the FFT result validates the kind setup, not this branch's ability to run FFT unaided.

## Known gaps versus GKE

Tabulated in dev/kind/README.md. The significant one is the single node —
cross-node scheduling and anything depending on workers landing apart cannot be
reproduced. Hardware is 24gb-tier only; the 80gb branch of tier selection is
covered by unit tests rather than here.

### CODE REVIEW

**Comment by droot** [comment]
@ShubyM @hodaaaaaaaaaa PTAL.

**Review by ShubyM — APPROVED** [review]

**Comment by droot** [comment]
closing this in favor of https://github.com/gke-labs/open-rl/pull/205
